KAFKA-14021: Implement new KIP-618 APIs in MirrorSourceConnector

C0urante opened the pull request:
[Jira](https://issues.apache.org/jira/browse/KAFKA-14021)

Implements the new `SourceConnector::exactlyOnceSupport` method in the source connector used by MirrorMaker 2. Since the connector tracks offsets using the Kafka Connect framework, exactly-once support is possible. However, we require that the consumer used to read from the source cluster is configured with the `read_committed` isolation level, since otherwise records from aborted and uncommitted transactions would be replicated.

Does not implement the new `SourceConnector::canDefineTransactionBoundaries` method as the default is to return `ConnectorTransactionBoundaries.UNSUPPORTED`, which is correct for this connector as it is incapable of defining its own transaction boundaries

A new integration test suite is also added for MirrorMaker 2 with exactly-once support enabled, with some small tweaks to run these tests correctly and account for changes in the Connect framework's behavior with exactly-once support.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

C0urante committed:
KAFKA-14021: Implement new KIP-618 APIs in MirrorSourceConnector

C0urante committed:
KAFKA-14021: Require MM2 source cluster consumers to use 'read_committed' isolation level for exactly-once support

C0urante commented:
@OmniaGM You've been active recently with MirrorMaker 2 and seem quite familiar with it. Would you be willing to give this a review? I'd love it if someone could fact-check me to make sure I've got the MM2-specific details here correct.

For reference:
- Docs for `SourceConnector::exactlyOnceSupport`: https://github.com/apache/kafka/blob/679e9e0cee67e7d3d2ece204a421ea7da31d73e9/connect/api/src/main/java/org/apache/kafka/connect/source/SourceConnector.java#L34-L54
- Docs and other useful info on the `exactly.once.support` connector property: https://github.com/apache/kafka/blob/679e9e0cee67e7d3d2ece204a421ea7da31d73e9/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SourceConnectorConfig.java#L60-L84

C0urante commented:
@tombentley If you have a moment, would you mind taking a look?

OmniaGM commented:
> @OmniaGM You've been active recently with MirrorMaker 2 and seem quite familiar with it. Would you be willing to give this a review? I'd love it if someone could fact-check here and make sure that the MM2-specific details are correct.
> 
> For reference:
> 
>     * Docs for `SourceConnector::exactlyOnceSupport`: https://github.com/apache/kafka/blob/679e9e0cee67e7d3d2ece204a421ea7da31d73e9/connect/api/src/main/java/org/apache/kafka/connect/source/SourceConnector.java#L34-L54
> 
>     * Docs and other useful info on the `exactly.once.support` connector property: https://github.com/apache/kafka/blob/679e9e0cee67e7d3d2ece204a421ea7da31d73e9/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SourceConnectorConfig.java#L60-L84

Hi sorry for the very late respond on this. I have the following question:
- What is the side effect of enabling `exactly.once.source.support=enabled` and `consumer.isolation.level=READ_COMMITTED` on the other `SourceConnector`s in MM2 like `MirrorHeartbeatConnector` and `MirrorCheckpointConnector` (both have no needs to be EOS). Will these also be EOS as well if the `MirrorSourceConnector` is configured to be EOS? If this is the case then KIP-618 need to mention this or/and update MM2 README.md.

C0urante commented:
Hi @OmniaGM--good questions. Due to [KAFKA-10586](https://issues.apache.org/jira/browse/KAFKA-10586), exactly-once isn't supported when running Mirror Maker 2 in dedicated mode (i.e., via `bin/connect-mirror-maker.sh`) and is instead only possible by deploying Mirror Maker 2 (or, specifically, one or more `MirrorSourceConnector` instances) onto a vanilla distributed Connect cluster.

Given that, I believe it's safe to say that the changes here should not require consideration of the heartbeat and checkpoint connectors. We may want to update the README to clarify the lack of support for exactly-once in dedicated mode, though.

Does that address your concerns?

OmniaGM commented:
Can we also add an integration test for EOS in MM2 as well? 

> Hi @OmniaGM -- good questions! Due to [KAFKA-10586](https://issues.apache.org/jira/browse/KAFKA-10586), exactly-once isn't supported when running Mirror Maker 2 in dedicated mode (i.e., via `bin/connect-mirror-maker.sh`) and is instead only possible by deploying Mirror Maker 2 (or, specifically, one or more `MirrorSourceConnector` instances) onto a vanilla distributed Connect cluster.
> 
> Given that, I believe it's safe to say that the changes here should not require consideration of the heartbeat and checkpoint connectors. We may want to update the README to clarify the lack of support for exactly-once in dedicated mode, though.
> 
> Does that address your concerns?

@C0urante Yeah updating the README to clear this should be enough.

C0urante committed:
Add integration test for exactly-once support in MirrorMaker 2, update README

C0urante commented:
Thanks @OmniaGM, good idea. I've updated the README and added an integration test that verifies that MM2 can still run with exactly-once support enabled.

I should note that the `testReplication` test case is currently failing locally with timeout issues, but succeeds when I bump the timeout. Going to see how the Jenkins build goes; we may choose to increase the timeout in these tests if they also fail during CI.

C0urante commented:
It turns out that the `testReplication` flakiness persisted in Jenkins, and was not solved by increasing timeouts.

Instead, the root of the problem was a change in the Connect framework's behavior when exactly-once support is enabled. Without exactly-once support, `SourceTask::commitRecord` is invoked [as soon as the record is ack'd](https://github.com/apache/kafka/blob/c1bb307a361b7b6e17261cc84ea2b108bacca84d/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java#L148) by the Kafka cluster, which usually causes those calls to be spread out over time. With exactly-once support, `SourceTask::commitRecord` is invoked [for every record in a transaction](https://github.com/apache/kafka/blob/c1bb307a361b7b6e17261cc84ea2b108bacca84d/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java#L312) once that transaction is committed, which causes a rapid series of calls to take place one after the other.

MirrorMaker 2 triggers a (potential) offset sync after every call to `commitRecord`, but it has [logic to prevent too many outstanding offset syncs from accruing](https://github.com/apache/kafka/blob/c1bb307a361b7b6e17261cc84ea2b108bacca84d/connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java#L208-L211). The exact limit on the number of outstanding offset requests [is ten](https://github.com/apache/kafka/blob/c1bb307a361b7b6e17261cc84ea2b108bacca84d/connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java#L53), which is less than the total number of topic partitions being replicated during the integration test. As a result, the test became flaky, since sometimes MM2 would drop an offset sync for partition 0 of the `test-topic-1` and then fail when [checking for offset syncs for that topic partition](https://github.com/apache/kafka/blob/c1bb307a361b7b6e17261cc84ea2b108bacca84d/connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationBaseTest.java#L319-L320).

Since the behavior change in the Connect framework may have an impact on MirrorMaker 2 outside of testing environments, I've tweaked the offset sync limit to apply on a per-topic-partition basis. This way, if a flurry of calls to `commitRecord` takes place when a transaction is committed, every topic partition should still get a chance for an offset sync, but there is still an upper bound on the number of outstanding offset syncs (although that bound is now proportional to the number of topic partitions being replicated, instead of a constant).

OmniaGM commented:
> Thanks @OmniaGM, good idea. I've updated the README and added an integration test that verifies that MM2 can still run with exactly-once support enabled.
> 
> I should note that the `testReplication` test case is currently failing locally with timeout issues, but succeeds when I bump the timeout. Going to see how the Jenkins build goes; we may choose to increase the timeout in these tests if they also fail during CI.

I had similar issue with the timeouts 

> It turns out that the `testReplication` flakiness persisted in Jenkins, and was not solved by increasing timeouts.
> 
> Instead, the root of the problem was a change in the Connect framework's behavior when exactly-once support is enabled. Without exactly-once support, `SourceTask::commitRecord` is invoked [as soon as the record is ack'd](https://github.com/apache/kafka/blob/c1bb307a361b7b6e17261cc84ea2b108bacca84d/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java#L148) by the Kafka cluster, which usually causes those calls to be spread out over time. With exactly-once support, `SourceTask::commitRecord` is invoked [for every record in a transaction](https://github.com/apache/kafka/blob/c1bb307a361b7b6e17261cc84ea2b108bacca84d/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java#L312) once that transaction is committed, which causes a rapid series of calls to take place one after the other.
> 
> MirrorMaker 2 triggers a (potential) offset sync after every call to `commitRecord`, but it has [logic to prevent too many outstanding offset syncs from accruing](https://github.com/apache/kafka/blob/c1bb307a361b7b6e17261cc84ea2b108bacca84d/connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java#L208-L211). The exact limit on the number of outstanding offset requests [is ten](https://github.com/apache/kafka/blob/c1bb307a361b7b6e17261cc84ea2b108bacca84d/connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java#L53), which is less than the total number of topic partitions being replicated during the integration test. As a result, the test became flaky, since sometimes MM2 would drop an offset sync for partition 0 of the `test-topic-1` and then fail when [checking for offset syncs for that topic partition](https://github.com/apache/kafka/blob/c1bb307a361b7b6e17261cc84ea2b108bacca84d/connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationBaseTest.java#L319-L320).
> 
> Since the behavior change in the Connect framework may have an impact on MirrorMaker 2 outside of testing environments, I've tweaked the offset sync limit to apply on a per-topic-partition basis. This way, if a flurry of calls to `commitRecord` takes place when a transaction is committed, every topic partition should still get a chance for an offset sync, but there is still an upper bound on the number of outstanding offset syncs (although that bound is now proportional to the number of topic partitions being replicated, instead of a constant).

Good finding. Should we add logs in `sendOffsetSync` to make it easier to find out if MirrorMaker is hitting the limit of `MAX_OUTSTANDING_OFFSET_SYNCS` in the future?

C0urante commented:
This was proposed in the original PR for MM2, but decided against: https://github.com/apache/kafka/pull/6295#discussion_r290372816

I was on the fence about adding something now that we're changing the logic. At `TRACE` level it's probably worth the effort; will add a log message.

mimaison commented on connect/mirror/README.md:
```diff
@@ -194,6 +194,10 @@ it is important to keep configuration consistent across flows to the same
 
+### Exactly-once
+Exactly-once delivery is not currently supported for dedicated MM2 clusters. In order to run MM2 with exactly-once, deploy it onto a
+Connect cluster that is configured with exactly-once support enabled. 
```
Should we explicitly say `Connect distributed cluster`?

C0urante replied:
👍 good call, will do.

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -69,6 +75,8 @@ public class MirrorSourceConnector extends SourceConnector {
         null, PatternType.ANY);
     private static final AclBindingFilter ANY_TOPIC_ACL = new AclBindingFilter(ANY_TOPIC, AccessControlEntryFilter.ANY);
+    private static final String READ_COMMITTED = IsolationLevel.READ_COMMITTED.toString().toLowerCase(Locale.ROOT);
+    private static final String EXACTLY_ONCE_SUPPORT_CONFIG = "exactly.once.support";
```
It's unfortunate we don't have this constant in the public API

C0urante replied:
Good point... probably can't add it without a KIP, though 🙁

mimaison replied:
Yeah we can't change this now but I thought I'd mention it.

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -185,11 +193,52 @@ public ConfigDef config() {
+                // We add a second error message here to give users more insight into why this specific connector can't support exactly-once
+                // guarantees with the given configuration
+                exactlyOnceSupport.addErrorMessage(
+                        "Mirror Maker 2 can only provide exactly-once guarantees when its source consumer is configured with "
```
I wonder if we should say `MirrorSourceConnector` instead of `Mirror Maker 2`.

C0urante replied:
Fine by me 👍

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -185,11 +193,52 @@ public ConfigDef config() {
+    private boolean consumerUsesReadCommitted(Map<String, String> props) {
+        MirrorConnectorConfig config = new MirrorConnectorConfig(props);
+        Object consumerIsolationLevel = config.sourceConsumerConfig().get(ConsumerConfig.ISOLATION_LEVEL_CONFIG);
+        return Objects.equals(READ_COMMITTED, consumerIsolationLevel);
```
Is `consumerIsolationLevel` guaranteed to be in lower case here?

C0urante replied:
It is not, but that's by design. The actual consumer property is case-sensitive; `READ_COMMITTED` is not accepted, but `read_committed` is.

mimaison replied:
Ah yes you're right! Thanks

mimaison reviewed (commented):
Thanks for the PR! I made a quick pass and left a few comments. Also I think this will need some updates following https://github.com/apache/kafka/pull/12899.

C0urante committed:
KAFKA-14021: Improve README wording and config validation error message

C0urante commented:
Thanks @mimaison. I've rebased and addressed your comments

mimaison commented:
@C0urante Let's try to get this in. If you rebase I'll take another look

C0urante commented:
@mimaison Thanks, I've rebased. I've also removed the changes to the offset sync semaphore since https://github.com/apache/kafka/pull/13181 should obviate them.

mimaison commented:
Thanks Chris for the quick reply! I'll take a look this week.

mimaison commented on connect/mirror/README.md:
```diff
@@ -194,6 +194,10 @@ it is important to keep configuration consistent across flows to the same
 target cluster. In most cases, your entire organization should use a single
 MM2 configuration file.
 
+### Exactly-once
```
This will change when https://github.com/apache/kafka/pull/13137 is merged

mimaison replied:
We merged https://github.com/apache/kafka/pull/13137 so this needs updating.

C0urante replied:
Yes, sorry for the delay here. I realized that the integration test added in that PR does not actually enable exactly-once on MM2 due a configuration error.

I've corrected the error in this PR, updated the MM2 README to include instructions on enabling exactly-once support, and added `isolation.level = read_committed` to the integration test for good measure to ensure that the steps we recommend in the README can actually be followed successfully.

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -188,11 +196,52 @@ public ConfigDef config() {
 
+    @Override
+    public org.apache.kafka.common.config.Config validate(Map<String, String> props) {
+        List<ConfigValue> configValues = MirrorSourceConfig.CONNECTOR_CONFIG_DEF.validate(props);
```
I wonder if we should do:
```
List<ConfigValue> configValues = super.validate(props).configValues();
```
so it's explicit that we perform the "regular" validation

C0urante replied:
Fine by me, done.

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -188,11 +196,52 @@ public ConfigDef config() {
+    }
+
+    private boolean consumerUsesReadCommitted(Map<String, String> props) {
+        MirrorSourceConfig config = new MirrorSourceConfig(props);
```
I think the issue is that if the configs contain a bad value this will throw. So when called from `validate()` we lose the error messages computed by `MirrorSourceConfig.CONNECTOR_CONFIG_DEF.validate(props);` and just throw `ConfigException`. 

You can reproduce if you add:
```
props.put(OFFSET_LAG_MAX, "bad");
configValue = validateProperty(OFFSET_LAG_MAX, props);
assertTrue(configValue.isPresent());
errorMessages = configValue.get().errorMessages();
assertEquals(1, errorMessages.size());
errorMessage = errorMessages.get(0);
assertTrue(
    errorMessages.get(0).contains(OFFSET_LAG_MAX),
    "Error message \"" + errorMessage + "\" should have mentioned the 'offset.lag.max' property"
);
```
to `MirrorSourceConnectorTest.testExactlyOnceSupportValidation()`

C0urante replied:
This turned out to be much trickier to address than expected, but I've made this method impervious to unrelated invalid properties.

This required being able to determine the full config used for the source consumer without constructing a `MirrorSourceConfig` instance, which in turn required (or at least, was facilitated by) pulling out the find-all-entries-with-prefix logic in `AbstractConfig::originalsWithPrefix` into a static method in `Utils`.

I couldn't think of a less invasive way to accomplish this, but it does seem worth the tradeoff; LMKWYT.

mimaison reviewed (commented):
Thanks for the update. I left a couple of questions about the `validate()` override in MirrorSourceConnector

C0urante committed:
Improve resilience of exactly-once support validation to invalid configs

C0urante committed:
Improve coverage for exactly-once support validation

mimaison reviewed (approved):
Thanks for the PR! Before merging can you update the support statement in the README?

C0urante committed:
Add docs on exactly-once support to MM2 README, fix broken integration test

C0urante commented:
@mimaison I've updated the README and fixed a small bug in the integration test recently introduced in https://github.com/apache/kafka/pull/13137. Mind giving one more quick pass?

mimaison commented on connect/mirror/README.md:
```diff
@@ -195,8 +195,46 @@ target cluster. In most cases, your entire organization should use a single
+recommended to secure the REST servers brought up by each MM2 node. See the configuration
+properties for Kafka Connect for information on how this can be accomplished.**
+
+It is also recommended to filter records from aborted transactions out from replicated data
```
Should `recommended` be `required`? as you can't enable EoS without setting `isolation.level=read_committed`

C0urante replied:
You technically can enable EOS without setting that property. It only becomes required if you also set `exactly.once.support` to `required` in the connector config, which has to be done explicitly (i.e., it isn't automatically even if you set `exactly.once.source.support` to `enabled` in your worker config).

Plus, it's actually impossible to set `exactly.once.support` for the `MirrorSourceConnector` in dedicated mode, since there's no way to specify a property for one type of Connector but not others, and others do not support it.

IMO it's better to use more lenient language here since it's not like you can't bring up MM2 with EOS enabled unless you set that specific property. If you don't find that convincing, we can change the wording.

mimaison reviewed (commented):
Thanks for the updates, I left a question. 

I also opened https://issues.apache.org/jira/browse/KAFKA-14709 as I think the MM2 docs should be in the actual docs and not in a README.

mimaison reviewed (approved):
LGTM, thanks!

C0urante commented:
Thanks Mickael!
